### ISSUE TITLE
Document guards in macro

### ISSUE THREAD

**Issue by stephanemagnenat** [lalrpop/lalrpop#973]
According to #921 and [this test](https://github.com/lalrpop/lalrpop/blob/d700700a5c32d31871067111b51ae86ed2a479b7/lalrpop/src/normalize/macro_expand/test.rs#L62-L73), lalrpop supports guards in macro. For certain use cases, such as when implementing a Rust-like language with optional trailing ";" in some control-flow expressions within blocks (as used in [this old partial Rust grammar implementation in lalrpop](https://github.com/nikomatsakis/rustypop/blob/master/src/rusty.lalrpop)), as the alternative is a lot of manual duplication.

I did not find any documentation of that feature, and as a user I think that it is worth documenting, hence this issue.

**Comment by dburgener** [comment]
Thanks for the rustypop link!  I looked into documenting this feature a few months ago, but I couldn't figure out what the motivating use-case was.  The rustypop example helps with that.  It's too involved to use directly in docs though, so I'll need to think on how to simplify it down.

Can you elaborate on the "optional trailing ";"" use case you mentioned?  This is basically the same contextual thing rustypop is doing where in some contexts you want to mandate a trailing ";" and some you don't, so you pass that context through a macro argument?  If you have any other lalrpop code to point me to, that would be helpful.  Like I said, rustypop is helpful, but in trying to distill the feature down to a good example for the documents, more inspiration around straightforward motivating use cases would be nice.

**Comment by stephanemagnenat** [comment]
> Can you elaborate on the "optional trailing ";"" use case you mentioned?

Yes. Rust allows to omit the trailing ";" when joining expressions in a block for certain expressions (called `ExpressionWithBlock` [here](https://doc.rust-lang.org/reference/expressions.html)). These expressions are essentially the ones that contain a mandatory block and therefore the trailing ";" is not strictly necessary to have a non-ambiguous grammar. However, while the resulting grammar is non-ambiguous, a naive implementation of it (e.g. e recursive tree of expressions and blocks) is, and some expressions should be disallowed in some contexts to avoid the ambiguity. The naive solution to this is to duplicate a large part of the grammar. The guard macro feature allows to avoid this, by keeping a single expression grammar "generic" over the macro parameters.

**Comment by yannham** [comment]
As a digression, I believe you can somtimes achieve the same result with standard LALRPOP macros as well. I used to make use of conditional macros as well, but I disliked the "untyped string" approach (if you make a spelling error, everything might compile fine but not run as you expect, for example). I realized I could replace it with normal macros, by parametrizing the _difference_ as a rule, instead of just a flag. For example:

```lalrpop
ContextDependantRule<C> : _ = {
  <Foo> "and" <Bar> => { ... },
  <Qux> "or" <Baz> => { ... },
  "if" <block> "else" <block> if C == "allow-else" => {...},
  "if" <block> if C == "disallow-else" => {...},
};
```

You can rewrite this by taking out the conditional as stand-alone rules, and parametrizing the rule by those other rules:

```lalrpop
ContextDependantRule<IfRule> : _ = {
  <Foo> "and" <Bar> => { ... },
  <Qux> "or" <Baz> => { ... },
  <IfRule> => <>,
};

IfElse : _ = "if" <if_block: Block> "else" <else_block: Block> => {...},
IfNoElse : _ = "if" <block: Block> => {...}
```

 It's slightly more verbose, but when applicable, at least you can't do stupid stuff like `ContextDependantRule<"allowelse">` which is incorrect won't be caught.

This doesn't work as nicely when instead of having a variation of one rule, you want to add rules conditionally (you can still do it but you need an empty "neutral" rule), or where several rules depend on the parameter, or more intricate examples as in the tests (several rules conditionally depend on the parameter).  I still suspect that it can be a viable alternative in many "dangling xxx" problems, where it doesn't add much boilerplate.

**Comment by stephanemagnenat** [comment]
@dburgener I create what I believe is the simplest reasonable example for this feature in #976.

@yannham interesting alternate idea!
